server/src/main/java/io/crate/expression/scalar/formatting/DateTimeFormatter.java

```diff
@@ -0,0 +1,503 @@
+
+        while (idx <= pattern_to_consume.length() && pattern_to_consume.length() > 0) {
+
+            if (idx == pattern_to_consume.length()) {
+                nextTokenNode = null;
+            } else {
+                nextTokenNode = currentTokenNode.children.get(pattern_to_consume.charAt(idx));
+            }
+
+            if (nextTokenNode == null && idx > 0) {
+                // No next step along the tree, so the parser should terminate and reset.
+                if (currentTokenNode.isTokenNode()) {
+                    // If the current node is a token, then a valid token is found and added to the list.
+                    // The token will then be removed from the rest of the string to parse.
+                    tokens.add(currentTokenNode.token);
+                } else {
+                    // If the current node is not a token, the parser then traverses back up the tree until it either
+                    // finds a token node, or the root.
+                    int depth = idx;
+
+                    while (depth > 0) {
+                        depth -= 1;
+                        currentTokenNode = currentTokenNode.parent;
+
+                        if (currentTokenNode.isTokenNode()) {
+                            // We have found the valid token, it should be added to the stack and the remaining rest
+                            // of the parsed pattern should be added as a string.
+                            tokens.add(currentTokenNode.token);
+                            break;
+                        }
+                    }
+
+                    tokens.add(pattern_to_consume.substring(depth, idx));
+                }
+
+                pattern_to_consume = pattern_to_consume.substring(idx);
+                currentTokenNode = ROOT_TOKEN_NODE;
+                idx = 0;
+            } else if (nextTokenNode == null) {
+                // If there is no path forward and index is 0, then we're at at the beginning, parsing a non-valid
+                // character. Add as a string.
+                tokens.add(pattern_to_consume.substring(0, idx + 1));
+                pattern_to_consume = pattern_to_consume.substring(idx + 1);
+            } else {
+                idx += 1;
+                currentTokenNode = nextTokenNode;
+            }
+        }
+
+        return tokens.toArray();
+    }
+
+    public String format(LocalDateTime datetime) {
+        String result = (String) Arrays.stream(this.tokens).sequential().reduce(
+            "", (str, elem) -> {
+                if (elem instanceof Token) {
+                    str = str + getElement((Token) elem, datetime);
+                } else if (elem instanceof String) {
+                    str = str + String.valueOf(elem);
+                }
+                return str;
+            }
+        );
+
+        return result;
+    }
+
+    private static String getElement(Token element, LocalDateTime datetime) {
+        switch (element) {
```

Comment:
If you want to play with newer language features - I think this would be suited for a switch expression: https://docs.oracle.com/en/java/javase/13/language/switch-expressions.html